docs/codeql/codeql-language-guides/customizing-library-models-for-python.rst

```diff
@@ -0,0 +1,449 @@
+
+.. code-block:: python
+
+  import invoke
+  c = invoke.Context()
+  c.run(cmd) # <-- add 'cmd' as a taint sink
+
+Note that this sink is already recognized by the CodeQL Python analysis, but for this example, you could use the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: sinkModel
+      data:
+        - ["invoke", "Member[Context].Instance.Member[run].Argument[0]", "command-injection"]
+
+- The first column, **"invoke"**, begins the search at places where the codebase imports the package **invoke**.
+- The second column is an access path that is evaluated from left to right, starting at the values that were identified by the first column.
+
+  - **Member[Context]** selects accesses to the **Context** class.
+  - **Instance** selects instances of the **Context** class.
+  - **Member[run]** selects accesses to the **run** method in the **Context** class.
+  - **Argument[0]** selects the first argument to calls to that method.
+
+- **command-injection** indicates that this is considered a sink for the command injection query.
+
+Note that the **Instance** component is used to select instances of a class, including instances of its subclasses.
+Since methods on instances are common targets, we have a more compact syntax for selecting them. The first column, the type, is allowed to contain a dotted path ending in a class name.
+This will begin the search at instances of that class. Using this syntax, the previous example could be written as:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: sinkModel
+      data:
+        - ["invoke.Context", "Member[run].Argument[0]", "command-injection"]
+
+Continued example: Multiple ways to obtain a type
+-------------------------------------------------
+
+The invoke package provides multiple ways to obtain a **Context** instance. The following example shows how to add a new way to obtain a **Context** instance:
+
+.. code-block:: python
+
+  from invoke import context
+  c = context.Context()
+  c.run(cmd) # <-- add 'cmd' as a taint sink
+
+Comparing to the previous Python snippet, the **Context** class is now found as **invoke.context.Context** instead of **invoke.Context**.
+We could add a data extension similar to the previous one, but with the type **invoke.context.Context**. However, we can also use the **typeModel** extensible predicate to describe how to reach **invoke.Context** from **invoke.context.Context**:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: typeModel
+      data:
+        - ["invoke.Context", "invoke", "Member[context].Member[Context].Instance"]
+
+- The first column, **"invoke.Context"**, is the name of the type to reach.
```

Comment:
```suggestion
- The first column, **"invoke.Context"**, is the name of the type that we're defining.
```
I found "type to reach" to be very vague and confusing. Perhaps this is a better formulation?

Reply:
I do not feel that we are defining this type, though. In cases where we make up a name, say to model some internal concept, I would say we are defining a type. But here, we are simply describing how one can obtain (an instance of) the type. Given that this can be used in various contexts, I actually like that it is a bit vague.